### PR TITLE
Timeout functions that read files in validation

### PR DESCRIPTION
**Pull request by emlys**
Addresses #261 

Using `threading` to set a timeout on functions that open a file. These could potentially be slowed down by Google File Stream-like services if the file has to be downloaded.

Note: I had a really hard time trying to replicate the issue with Google File Stream. Using a 2GB file on my google drive, it seemed to stay available locally no matter what I did. I tried limiting the download speed using the File Stream setting to be sure it would take a noticeable amount of time to download. Even after uninstalling, clearing the cache, and reinstalling, the file would show up almost instantly.

### CODE REVIEW

**Comment by phargogh** [comment]
FYI @emlys I just changed the target branch to `main`, which looks to have been pretty seamless!

File: src/natcap/invest/validation.py
Comment (emlys) [inline src/natcap/invest/validation.py:631]: @phargogh and I discussed that it might be better to raise an error on timeout and exit validation entirely. I first tried it out this way, just returning the timeout message like any other validation warning message, because it took the least modification to things outside of `timeout()`. But this could maybe be a good use for a custom exception?

File: src/natcap/invest/validation.py
Comment (dcdenu4) [inline src/natcap/invest/validation.py:529]: Was this a linting change?

File: src/natcap/invest/validation.py
Comment (dcdenu4) [inline src/natcap/invest/validation.py:617]: extra line here?

File: src/natcap/invest/validation.py
Comment (dcdenu4) [reply src/natcap/invest/validation.py:631]: I also thing there's a `TimeoutError` that could be appropriate as opposed to a custom exception.

Comment (dcdenu4) [reply src/natcap/invest/validation.py:631]: I do like the idea of not holding the user hostage if this timeouts. Meaning, on a given timeout I as the user should still be able to click run. Is that what happens in this case?

File: src/natcap/invest/validation.py
Comment (dcdenu4) [inline src/natcap/invest/validation.py:641]: Should this comment be: 
>opening a file could potentially take a long time

File: src/natcap/invest/validation.py
Comment (dcdenu4) [inline src/natcap/invest/validation.py:806]: Was this a linting suggestion as well? Just curious.

File: tests/test_validation.py
Comment (dcdenu4) [inline tests/test_validation.py:728]: Neat. I haven' done much Mocking. Would that line behave the same as:
```python
with unittest.mock.patch('natcap.invest.validation.check_csv', mock_read_csv)
```
?

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py:631]: It would still block you from running the model, that input would show up with a red X. That is interesting though, should it be possible to run the model anyway?

I kind of like this approach because it ensures that the UI will show up. Apparently if there are saved inputs from a previous run, validation happens before the model window opens. So if one of the inputs needs to be downloaded it could appear to hang while that happens. I like that this way the error message shows up in the UI. If it raised and exception instead it would have to show up in the terminal window instead and the user might not notice it.

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py:529]: no I did that when I was going to use an exception instead of warning message. Forgot to change it back 🙃

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py:806]: oops! no this was just left over from a previous change

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py]: Thanks for catching that! Rephrased it to make more sense.

File: tests/test_validation.py
Comment (emlys) [reply tests/test_validation.py:728]: I think that would work also... to be consistent I'd need to change the `Mock` return value to be a list, not a dataframe. But yes that should also trigger the timeout. Do you think it would be better that way? I guess it decreases the amount of irrelevant code that's run in the test

**Comment by emlys** [comment]
Adding @phargogh because we had talked about this issue

File: tests/test_validation.py
Comment (dcdenu4) [reply tests/test_validation.py:733]: Hmm, I'm not sure about "better" but I'll leave it up to you. Maybe your last comment of decreasing the amount of code that's run makes sense? I guess `pandas.read_csv` is the actual function that would take awhile to load, but to me `check_csv` is a more clear function to mock... I guess 6 in one hand, half dozen in the other.

**Review by dcdenu4 — CHANGES_REQUESTED** [review]
Hey @emlys thanks for these comment and changes. I just had one more comment about raising a `RuntimeError`.

File: src/natcap/invest/validation.py
Comment (dcdenu4) [reply src/natcap/invest/validation.py:618]: Oh, oops. I thought this said `RuntimeError`. I see. Please disregard.

**Review by dcdenu4 — APPROVED** [review]

File: src/natcap/invest/validation.py
Comment (phargogh) [reply src/natcap/invest/validation.py]: For future us, we were chatting about this on our internal slack channel and decided that an inability to run validation shouldn't prevent an informed user from running a model.  Rather, we'll want to display some sort of warning about the inability to validate and let the user make an informed decision.

In the longer-run, this will involve reworking the validation API to support status levels.

File: HISTORY.rst
Comment (phargogh) [inline HISTORY.rst:41]: The original issue #261 has to do with file streaming services, but the approach we're taking here is more general and would apply to any file-based validation that takes too long.

And would it be worth mentioning that it'll raise a `warning` rather than an error?

File: src/natcap/invest/validation.py
Comment (phargogh) [inline src/natcap/invest/validation.py:658]: Neat use of `functools.partial`!

Do you think it'd be useful to also wrap `file`, `folder` and `directory` with the timeout function as well?  Since the idea of the new `timeout` function here is to have a timeout on file-based operations and these three also make OS calls to check if a file exists, I'm wondering if we're likely to get into a case where even checking to see if a file or folder exists on disk could take some time, like if someone is checking an NFS share or something like that.

As a separate note about the use of `functools.partial`, wrapping the timeout functions in this way unfortunately does not allow for the timeout to differ per use case.  Here are a few that come to mind that might benefit from different timeouts:

1. The Qt UI - here I think the default 5s timeout could work just fine ... we're just trying to prevent a feeling of unresponsiveness.
2. The workbench - we've previously talked about timeouts in the context of that it would be great to be able to basically guarantee fast validation ... so in that case, maybe we'd want validation to return within a much shorter timeframe.  In any case, I suspect it would be useful to be able to configure the timeout so that the workbench and/or server can set their own limits as needed.
3. Model execution - there are a number of model that call `validate` within the model's execute.  In this case it could be reasonable to permit a longer-running validation.

All that being said, perhaps this isn't the right place to address these timeout use cases, and the solution as stated _does_ address the issue at hand so maybe it'd be worth addressing this along with the other design issues in the validation API that came up on the slack thread yesterday.  What do you think @dcdenu4 @emlys ?

File: tests/test_validation.py
Comment (phargogh) [inline tests/test_validation.py:230]: Do you think it would be reasonable to trim down the sleep time here?  10s isn't a terribly long time to wait in the scope of the rest of the test suite, but it also doesn't seem like that long of a wait is required to test the timeout function.

File: tests/test_validation.py
Comment (phargogh) [inline tests/test_validation.py:711]: Same here ... between the `time.sleep` and the timeout we can set, would it be reasonable to turn down the expected runtime of these tests?  Just hoping to have our tests run as quickly as is reasonable, is all!

File: tests/test_validation.py
Comment (phargogh) [inline tests/test_validation.py:736]: This is pretty minor since we don't use warnings all that much throughout InVEST, but this looks to me like installing this filter is altering a global `warnings` state ... is that the case?  If so, would it be possible to uninstall the filter once the test completes, on both test success and test failure so we can avoid the side effect?  Something like

```python
try:
    warnings.simplefilter("always")
    validation.validate(args, spec)
    self.assertTrue(len(ws) == 1)
    self.assertTrue('timed out' in str(ws[0].message))
finally:
    <uninstall the warnings filter>
```

File: src/natcap/invest/validation.py
Comment (phargogh) [inline src/natcap/invest/validation.py:13]: Is signal unused?  Maybe this was left over from a prototype?

**Review by phargogh — CHANGES_REQUESTED** [review]
Hey @emlys this looks like a nice solution to the file streaming issue!

I had a couple of pretty minor requests, and then one larger question about the nature of a timeout in the scope of a couple of different use cases we have that I think would be worth talking about at least a little and absolutely might be outside of the scope of this PR.,  But I wanted to bring it up and I'd love to hear what you think!

File: tests/test_validation.py
Comment (emlys) [reply tests/test_validation.py:733]: Happy to change it!

File: tests/test_validation.py
Comment (emlys) [reply tests/test_validation.py:230]: Yes definitely, 6 seconds should be fine

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py:13]: oops, forgot to take that out!

File: tests/test_validation.py
Comment (emlys) [reply tests/test_validation.py:736]: Looks like it should be restored to the original state after exiting the context manager: https://docs.python.org/3/library/warnings.html#testing-warnings

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py:658]: It could definitely be useful to have a timeout for `file`, `folder`, and `directory` as well! I didn't include them at first because I'm pretty sure that doing `os.path.exists` on a file stream filepath wouldn't trigger it to download, but there are probably other situations where it could take a long time, so I'll add the timeouts there.

It should work to pass in a different time to `timeout`...
`functools.partial(timeout, check_csv, timeout=10)` --> `timeout(check_csv, *args, timeout=10, **kwargs)`, and then the thread target becomes `check_csv(*args, **kwargs)`. Changing the timeout this way does seem to work in the tests, so I'm pretty confident, but it's definitely not very intuitive how the args get passed around

Comment (phargogh) [reply src/natcap/invest/validation.py:658]: Ok great!  Yes, probably safest to add timeouts to all of the filesystem operations in case those take some time.

For the timeouts, it makes sense to pass the timeout argument around, and after some reflection, I think I might suggest that we wait to address that until we have a more concrete plan ironed out for a round of updates to validation.  What do you think?

File: tests/test_validation.py
Comment (emlys) [reply tests/test_validation.py]: Turns out that it's not so simple to replace `'natcap.invest.validation.pandas.read_csv'` with `'natcap.invest.validation.check_csv'` because `validate` references `check_csv` via `functools.partial` in the `_VALIDATE_FUNCS` dictionary. Ended up mocking the dictionary instead and replacing `functools.partial(timeout, check_csv)` with `functools.partial(timeout, delay)` :)

File: src/natcap/invest/validation.py
Comment (emlys) [reply src/natcap/invest/validation.py:657]: Agree that we can wait on adding that option until we plan to use it for those different cases 👍

File: HISTORY.rst
Comment (emlys) [reply HISTORY.rst:41]: good point!

File: tests/test_validation.py
Comment (phargogh) [reply tests/test_validation.py:735]: Huh, so it does!  Thanks for looking into that.

**Comment by dcdenu4** [comment]
@phargogh @emlys I'm in agreement that it's fine to hold off on some of the `timeout` implementation and consider it's use in a larger redesign of the validation module.

**Review by phargogh — APPROVED** [review]
Great, thanks @emlys !

**Pull request merged**
